### PR TITLE
S4143 False positives with `List<T>.Add` method calls

### LINKED ISSUE #2674: S4143 False positives with `List<T>.Add` method calls
### Description

Using the SonarAnalyzer.CSharp nuget, S4143 flags issues for List<String> (and probably any lists)

### Repro steps

Using the code:
```
List<string> S4143 = new List<string>();
S4143.Add("SomeText");
S4143.Add("\n");
S4143.Add("SomeText");
S4143.Add("\n");
```

The last 2 Adds are showing with warnings for S4143, as they are Adds they're extending the list rather than overwriting any existing entries.

### Expected behavior

The above code shouldn't generate a warning.

### Actual behavior

S4143 generates warnings.

### Known workarounds

Set the warning to info or suppress.

### Related information

* SonarC# Version: 7.16.0.8981
* Visual Studio Version: 2017, 15.9.16
* If running through the Scanner for MSBuild, its version

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #2674

### CODE REVIEW

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotOverwriteCollectionElements.vb [hunk 1]
```diff
@@ -4,214 +4,232 @@ Imports System.Linq
...
+            dict.add(0, 1) ' Noncompliant
+'           ^^^^^^^^^^^^^^
+        End Sub
+
+        Private Sub IDictionaryAddMixedCase(ByVal dict As IDictionary(Of Integer, Integer))
+            dict.Add(0, 0) ' Secondary
+'           ^^^^^^^^^^^^^^
+            dict.Add(0, 1) ' Noncompliant
+'           ^^^^^^^^^^^^^^
+        End Sub
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotOverwriteCollectionElements.vb]: This test used to check whether the rule would also work when casing is different than CamelCase. The formatting here removes the original purpose of the test. It should remain as it was originally:
```
      dict.add(0, 0) ' Secondary
'     ^^^^^^^^^^^^^^
      dict.ADD(0, 1) ' Noncompliant
'     ^^^^^^^^^^^^^^
```

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotOverwriteCollectionElements.vb [hunk 2]
```diff
@@ -4,214 +4,232 @@ Imports System.Linq
...
+
+    End Class
 
     'See https://github.com/SonarSource/sonar-dotnet/issues/2235
     Class NullReferenceReproducer
+
         Sub FooBar()
-            Bar ' AD0001 NullReferenceException in GetFirstArgumentExpression
+            Bar() ' AD0001 NullReferenceException in GetFirstArgumentExpression
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotOverwriteCollectionElements.vb]: Formatting also removed the purpose of the test here. The method call without parenthesis was added to make sure that rule does not throw `NullReferenceException` when method invocation is done without argument. See PR #2237

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotOverwriteCollectionElements.vb (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotOverwriteCollectionElements.vb]: I've actually fix this already, but haven't noticed that I got autoformatted again :(

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotOverwriteCollectionElements.vb (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotOverwriteCollectionElements.vb]: Same as above

**Pull request merged**